src/mlpack/core/data/load_impl.hpp

```diff
@@ -20,184 +21,233 @@
+          arma::csv_opts::trans + arma::csv_opts::semicolon), arma::csv_ascii);
 
-  // We can't use the stream if the type is HDF5.
-  bool success;
-  LoadCSV loader;
+  else if (opts.NoTranspose() && opts.HasHeaders() &&
+           opts.SemiColon() && !opts.MissingToNan())
 
-  if (loadType != FileType::HDF5Binary)
-  {
-    if (loadType == FileType::CSVASCII)
-      success = loader.LoadNumericCSV(matrix, stream);
-    else
-      success = matrix.load(stream, ToArmaFileType(loadType));
-  }
-  else
-    success = matrix.load(filename, ToArmaFileType(loadType));
+    success = matrix.load(arma::csv_name(filename, opts.Headers(),
+          arma::csv_opts::semicolon), arma::csv_ascii);
 
-  if (!success)
-  {
-    Log::Info << std::endl;
-    Timer::Stop("loading_data");
-    if (fatal)
-      Log::Fatal << "Loading from '" << filename << "' failed." << std::endl;
-    else
-      Log::Warn << "Loading from '" << filename << "' failed." << std::endl;
+  else if (opts.NoTranspose() && opts.HasHeaders() &&
+           !opts.SemiColon() && opts.MissingToNan())
 
-    return false;
-  }
-  else
-    Log::Info << "Size is " << (transpose ? matrix.n_cols : matrix.n_rows)
-        << " x " << (transpose ? matrix.n_rows : matrix.n_cols) << ".\n";
+#ifdef NOT_MISSING_TO_NAN
+    success = matrix.load(arma::csv_name(filename, opts.Headers()),
+        arma::csv_ascii);
+#else
+    success = matrix.load(arma::csv_name(filename, opts.Headers(),
+          arma::csv_opts::strict), arma::csv_ascii);
+#endif
+    
+  else if (opts.NoTranspose() && opts.HasHeaders() &&
+            opts.SemiColon() && opts.MissingToNan())
 
-  // Now transpose the matrix, if necessary.
-  if (transpose)
-  {
-    success = inplace_transpose(matrix, fatal);
-  }
+#ifdef NOT_MISSING_TO_NAN
+    success = matrix.load(arma::csv_name(filename, opts.Headers(),
+        arma::csv_opts::semicolon), arma::csv_ascii);
+#else
+    success = matrix.load(arma::csv_name(filename, opts.Headers(),
+        arma::csv_opts::strict + arma::csv_opts::semicolon), arma::csv_ascii);
+#endif
 
-  Timer::Stop("loading_data");
+  else if (!opts.NoTranspose() && !opts.HasHeaders() &&
+           opts.SemiColon() && opts.MissingToNan())
+
+#ifdef NOT_MISSING_TO_NAN
+    success = matrix.load(arma::csv_name(filename, arma::csv_opts::trans +
+        arma::csv_opts::semicolon), arma::csv_ascii);
+#else
+    success = matrix.load(arma::csv_name(filename, arma::csv_opts::strict +
+          arma::csv_opts::trans + arma::csv_opts::semicolon), arma::csv_ascii);
+#endif
+
+  else if (opts.NoTranspose() && !opts.HasHeaders() &&
+           opts.SemiColon() && opts.MissingToNan())
+
+#ifdef NOT_MISSING_TO_NAN
+    success = matrix.load(arma::csv_name(filename, arma::csv_opts::semicolon),
+        arma::csv_ascii);
+#else
+    success = matrix.load(arma::csv_name(filename, arma::csv_opts::strict +
+          arma::csv_opts::semicolon), arma::csv_ascii);
+#endif
```

Comment:
It should be possible to avoid this combinatoric explosion of options here.  If we are using the Armadillo CSV loader, then it may make sense to structure something like this...

```
arma::csv_name c(filename);
if (strict)
  c += arma::csv_opts::strict;
if (semicolon)
  c += arma::csv_opts::semicolon;
if (!noTranspose)
  c += arma::csv_opts::trans;
```

and so on and so forth, you get the idea.